command/agent/syslog.go

```diff
@@ -63,3 +73,48 @@ func (s *SyslogWrapper) Write(p []byte) (int, error) {
 	err := s.l.WriteLevel(getSysLogPriority(level), afterLevel)
 	return len(p), err
 }
+
+var (
+	// jsonLogLineLevelRegex is used to find the log level key/value entry
+	// within a JSON log line. It will match string entries such as
+	// `"@level":"debug",`, so we can pull these out for syslog capabilities.
+	jsonLogLineLevelRegex = regexp.MustCompile(`"@level":"\w+",`)
+)
+
+// syslogJSONWrapper is a syslog writer for Nomad logs when the operator has
+// enabled the JSON logging format.
+type syslogJSONWrapper struct {
+	logger gsyslog.Syslogger
+}
+
+// Write is used to implement io.Writer. It dissects the passed JSON log line,
+// identifying the log level and removing the contextual entry, before
+// performing the syslog write.
+//
+// Nomad's syslog is fed by go-hclog which is responsible for performing the
+// log level filtering. It is not needed here.
+func (s *syslogJSONWrapper) Write(logBytes []byte) (int, error) {
+
+	// Find the start and finish index of the regex match, so we know where in
+	// the byte array the level contextual entry is.
+	indexes := jsonLogLineLevelRegex.FindAllIndex(logBytes, 1)
+
+	// If the indexes are not what we expected, write the log line with the
+	// notice level. It's better to have a log line at the incorrect level
+	// than have a log line saying we couldn't write a log line.
+	if len(indexes) != 1 || len(indexes[0]) != 2 {
+		return len(logBytes), s.logger.WriteLevel(gsyslog.LOG_NOTICE, logBytes)
+	}
+
+	// Pull the log level from the message using the identified indexes and
+	// knowledge of the JSON formatting from go-hclog.
+	level := strings.ToTitle(string(logBytes[indexes[0][0]+10 : indexes[0][1]-2]))
+
+	// Rip the level out of the log message line. The level will be indicated
+	// by syslog, depending on the operators formatting setup.
+	afterLevel := logBytes[:indexes[0][0]]
+	afterLevel = append(afterLevel, logBytes[indexes[0][1]:]...)
```

Comment:
This seems like an expensive bit of extra string mangling to do. Do we really need to remove the original `@level`?

Reply:
Done within: https://github.com/hashicorp/nomad/pull/24865/commits/eb32515129074c6cf843c923009c67fe958425f2

Reply:
I can see a case where it might be confusing to have two different levels.

But I _think_ that because we're doing this before we're calling `WriteLevel`, we have to do this work for _all_ logs. Ex. we're doing this work for trace, debug, and info level and then throwing it out if you're logging at warn. (Unless the level gate is earlier than I'm seeing here, in which case maybe it doesn't matter.)

Reply:
The level gate does come earlier as these logs are fed from hclog which performs that work. Any logs coming in here, are expected to be written.

I think the two levels are easy enough to reason about it shouldn't be too confusing. Depending on the syslog configuration, the priority might not even be included in the syslog entry output.